File: SmartReceipts/Backup/DataImport.m
//
//  DataImport.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 27/05/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "DataImport.h"
#import "Objective-Zip.h"
#import "Constants.h"
#import "WBFileManager.h"
#import "WBPreferences.h"

@interface DataImport ()

@property (nonatomic, copy) NSString *inputPath;
@property (nonatomic, copy) NSString *outputPath;

@end

@implementation DataImport

- (id)initWithInputFile:(NSString *)inputPath output:(NSString *)outputPath {
    self = [self init];
    if (self) {
        _inputPath = inputPath;
        _outputPath = outputPath;
    }
    return self;
}

- (void)execute {
    NSError *error;
    BOOL result = [[NSFileManager defaultManager] createDirectoryAtPath:self.outputPath withIntermediateDirectories:YES attributes:nil error:&error];
    if ((result == NO) || error != nil) {
        LOGGER_ERROR(@"execute(), error creating directory at path: %@, error: %@", self.outputPath, error.description);
    }
    
    OZZipFile *zipFile = [[OZZipFile alloc] initWithFileName:self.inputPath mode:OZZipFileModeUnzip];
    [self extractFromZip:zipFile zipName:SmartReceiptsDatabaseExportName toFile:[self.outputPath stringByAppendingPathComponent:SmartReceiptsDatabaseExportName]];
    NSData *preferences = [self extractDataFromZip:zipFile withName:SmartReceiptsPreferencesExportName];
    [WBPreferences setFromXmlString:[[NSString alloc] initWithData:preferences encoding:NSUTF8StringEncoding]];

    // trips contents
    [zipFile goToFirstFileInZip];
    do {
        OZFileInZipInfo *info = [zipFile getCurrentFileInZipInfo];
        NSString *name = info.name;
        if ([name isEqualToString:SmartReceiptsDatabaseExportName]) {
            continue;
        }
        if ([name hasPrefix:@"shared_prefs/"]) {
            continue;
        }

        NSArray *components = [name pathComponents];
        if (components.count != 2) {
            continue;
        }

        NSString *tripName = components[0];
        NSString *fileName = components[1];

        LOGGER_INFO(@"Extract file for trip:%@", tripName);
        NSString *tripPath = [[self.outputPath stringByAppendingPathComponent:SmartReceiptsTripsDirectoryName] stringByAppendingPathComponent:tripName];
        [[NSFileManager defaultManager] createDirectoryAtPath:tripPath withIntermediateDirectories:YES attributes:nil error:nil];
        NSString *filePath = [tripPath stringByAppendingPathComponent:fileName];
        OZZipReadStream *stream = [zipFile readCurrentFileInZip];
        [self writeDataFromStream:stream toFile:filePath];
    } while ([zipFile goToNextFileInZip]);
}

- (NSData *)extractDataFromZip:(OZZipFile *)zipFile withName:(NSString *)fileName {
    NSString *tempFile = [NSTemporaryDirectory() stringByAppendingPathComponent:@"extract"];
    [self extractFromZip:zipFile zipName:fileName toFile:tempFile];
    NSData *data = [NSData dataWithContentsOfFile:tempFile];
    [[NSFileManager defaultManager] removeItemAtPath:tempFile error:nil];
    return data;
}

- (void)extractFromZip:(OZZipFile *)zipFile zipName:(NSString *)zipName toFile:(NSString *)outPath {
    LOGGER_DEBUG(@"Extract file named: %@", zipName);
    BOOL found = [zipFile locateFileInZip:zipName];
    if (!found) {
        LOGGER_WARNING(@"File with name %@ not in zip", zipName);
        return;
    }

    OZZipReadStream *stream = [zipFile readCurrentFileInZip];
    [self writeDataFromStream:stream toFile:outPath];
}

- (void)writeDataFromStream:(OZZipReadStream *)stream toFile:(NSString *)file {
    @autoreleasepool {
        NSMutableData *buffer = [[NSMutableData alloc] initWithLength:(8 * 1024)];
        NSMutableData *resultData = [[NSMutableData alloc] init];
        NSUInteger len;
        while ((len = [stream readDataWithBuffer:buffer]) > 0) {
            [resultData appendBytes:[buffer mutableBytes] length:len];
        }
        [stream finishedReading];

        LOGGER_DEBUG(@"File size %tu", resultData.length);
        [WBFileManager forceWriteData:resultData to:file];
        LOGGER_DEBUG(@"Written to %@", file);
    }
}

@end

File: SmartReceipts/Backup/DataImport.swift
//
//  DataImport.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 11/06/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import objective_zip

class DataImport: NSObject {
    
    var inputPath: String!
    var outputPath: String!
    
    init(inputFile: String, output: String) {
        super.init()
        self.inputPath = inputFile
        self.outputPath = output
    }
    
    func execute() {
        do {
            try FileManager.default.createDirectory(atPath: outputPath, withIntermediateDirectories: true, attributes: nil)
        } catch {
            Logger.error("execute(), error creating directory at path: \(outputPath), error: \(error.localizedDescription)")
        }
        let zipFile = OZZipFile(fileName: inputPath, mode: .unzip)
        let toFile = (outputPath as NSString).appendingPathComponent(SmartReceiptsDatabaseExportName)
        extractFrom(zip: zipFile, zipName: SmartReceiptsDatabaseExportName, toFile: toFile)
        let preferences = extractDataFrom(zipFile: zipFile, fileName: SmartReceiptsPreferencesExportName)
        WBPreferences.setFromXmlString(String(data: preferences!, encoding: .utf8))
        
        // Trips contents
        zipFile.goToFirstFileInZip()
        
        repeat {
            let info = zipFile.getCurrentFileInZipInfo()
            let name = info.name
            if name == SmartReceiptsDatabaseExportName || name.hasPrefix("shared_prefs/") {
                continue
            }
            
            let components = (name as NSString).pathComponents
            if components.count != 2 {
                continue
            }
            
            let tripName = components[0]
            let fileName = components[1]
            
            Logger.debug("Extract file for trip: \(tripName)")
            let tripPath = ((outputPath as NSString).appendingPathComponent(SmartReceiptsTripsDirectoryName)
                as NSString).appendingPathComponent(tripName)
            try? FileManager.default.createDirectory(atPath: tripPath, withIntermediateDirectories: true, attributes: nil)
            let filePath = (tripPath as NSString).appendingPathComponent(fileName)
            let stream = zipFile.readCurrentFileInZip()
            writeDataFrom(stream: stream, toFile: filePath)
        } while zipFile.goToNextFileInZip()
    }
    
    func extractFrom(zip: OZZipFile, zipName: String, toFile: String) {
        Logger.debug("Extract file named: \(zipName)")
        let found = zip.locateFile(inZip: zipName)
        if (!found) {
            Logger.warning("File with name \(zipName) not in zip")
            return;
        }
        let stream = zip.readCurrentFileInZip()
        writeDataFrom(stream: stream, toFile: toFile)
    }
    
    func writeDataFrom(stream: OZZipReadStream, toFile: String) {
        let buffer = NSMutableData(length: 8 * 1024)!
        let resultData = NSMutableData()
        var len = stream.readData(withBuffer: buffer)
        while len > 0 {
            resultData.append(buffer.mutableBytes, length: Int(len))
            len = stream.readData(withBuffer: buffer)
        }
        stream.finishedReading()
        
        Logger.debug("File size \(resultData.length)")
        WBFileManager.forceWrite(resultData as Data, to: toFile)
        Logger.debug("Written to \(toFile)")
    }
    
    func extractDataFrom(zipFile: OZZipFile, fileName: String) -> Data? {
        let tempFile = (NSTemporaryDirectory() as NSString).appendingPathComponent("extract")
        extractFrom(zip: zipFile, zipName: fileName, toFile: tempFile)
        let data = NSData(contentsOfFile: tempFile)
        try? FileManager.default.removeItem(atPath: tempFile)
        return data as Data?
    }
}
